src/components/data/QueryAuth0.js

```diff
@@ -38,100 +40,117 @@ export default class QueryAuth0 extends React.PureComponent {
-      }
-      // But this is only good enough if it hasn't expired.
-      const expiresAt = JSON.parse(localStorage.getItem('expiresAt'));
-      if (expiresAt && expiresAt - new Date().getTime() < 0) {
-        // Oh no! It has expired.
-        authResult = null;
+      expiresAt = JSON.parse(localStorage.getItem('expiresAt'));
+      if (expiresAt && expiresAt - new Date().getTime() > 0) {
+        authResult = JSON.parse(localStorage.getItem('authResult'));
       }
     }
 
     if (authResult) {
-      // This is only true if you arrived on the site with a valid window.location.hash
-      // or the localStorage.authResult (and localStorage.expiresAt) wasn't out of date.
-      this.postProcessAuthResult(authResult);
-    }
-    if (startAccessTokenRefreshLoop) {
-      // Start a never-ending loop of periodically checking if the accessToken is about to
-      // or has expired.
-      // This only happens for users who have successfully logged in or have successfully
-      // logged in the *past* but expired.
-      await this.accessTokenRefreshLoop();
+      // If we have a valid authResult log the user in
+      this.processAuthResult(authResult);
+    } else if (expiresAt && expiresAt - new Date().getTime() < 0) {
+      // If we do not have an expired authResult attempt to refresh it
+      this.refreshAccessToken();
+    } else if (!authResult && accessToken) {
+      this.props.logUserOut();
     }
   }
 
-  postProcessAuthResult = authResult => {
+  processAuthResult(authResult) {
     const { logUserIn, userProfileReceived } = this.props;
-    const { state } = authResult;
+    const { state: redirectUrl } = authResult;
+
     logUserIn(authResult);
-    // Since we include 'id_token' for the 'responseType' in auth0.WebAuth
-    // the authresult will contain the user profile as .idTokenPayload
-    // included with the accessToken. Use this now to update the state so that we
-    // can display the name and avatar you logged in as.
     userProfileReceived(authResult.idTokenPayload);
 
-    if (state) {
-      this.props.push(state);
+    if (redirectUrl) {
+      this.props.push(redirectUrl);
     }
-  };
+  }
 
-  async accessTokenRefreshLoop() {
-    if (!localStorage.getItem('expiresAt')) {
-      // The user has logged out probably. Either way, no point bothering to refresh.
-      return;
+  validateAccessToken() {
+    const { accessToken } = this.props;
+    if (accessToken) {
+      const expiresAt = JSON.parse(localStorage.getItem('expiresAt'));
+      const expiresIn = expiresAt - new Date().getTime();
+
+      if (expiresIn / 1000 <= CHECK_AUTH_EXPIRY_INTERVAL_SECONDS) {
+        // The token will expire before we check again so attempt a refresh now
+        this.refreshAccessToken();
+      }
     }
-    const expiresAt = JSON.parse(localStorage.getItem('expiresAt'));
-    const expiresIn = expiresAt - new Date().getTime();
-    const shouldRefresh = expiresIn <= CHECK_AUTH_EXPIRY_INTERVAL_SECONDS;
```

Comment:
This appears to be the culprit for the first case. If it's 1m till expiry `expiresIn`=60000. Meanwhile `CHECK_AUTH_EXPIRY_INTERVAL_SECONDS`=300. It was a units issue.

Reply:
Gross! 
I feel guilty because you gave me that line in your PR feedback. Your intent was just to structure the code in a more neat way and I just copied and pasted it in without noticing the flaw.

Reply:
I should have realized in my example too. Oh well! I actually thought about this some more and since JS usually handles time in milliseconds we should have that be the unit of the setting.